pkg/model/diamond.go

```diff
@@ -0,0 +1,265 @@
+	EnableConflicts ConflictMode = "enable-conflicts"
+
+	// ForbidConflicts is the diamond mode with which conflicts result in failure to commit
+	ForbidConflicts ConflictMode = "forbids-conflicts"
+)
+
+// DiamondState models the running status of an ungoing diamond workflow
+type DiamondState string
+
+const (
+	// DiamondInitialized is the state of an initialized diamond
+	DiamondInitialized DiamondState = "initialized"
+
+	// DiamondCommitting indicates that a commit is going on for this diamond
+	DiamondCommitting DiamondState = "committing"
+
+	// DiamondCancelling indicates that a cancel is going on for this diamond
+	DiamondCancelling DiamondState = "cancelling"
+
+	// DiamondRunning indicates that the diamond has started some splits
+	DiamondRunning DiamondState = "running"
+
+	// DiamondDone indicates the diamond has completed with a commit
+	DiamondDone DiamondState = "done"
+
+	// DiamondCancelled indicates the diamond has completed with a cancel
+	DiamondCancelled DiamondState = "cancelled"
+)
+
+const (
+	diamondDescriptorFile = "diamond.yaml"
+	splitDescriptorFile   = "split.yaml"
+)
+
+// IsValid checks the value of a diamond state
+func (s DiamondState) IsValid() bool {
+	switch s {
+	case DiamondInitialized, DiamondRunning, DiamondCommitting, DiamondDone, DiamondCancelling, DiamondCancelled:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s DiamondState) String() string {
+	return string(s)
+}
+
+// SplitState models the running status of an ungoing diamond split
+type SplitState string
+
+const (
+	// SplitDone is the state of a completed split
+	SplitDone SplitState = "done"
+
+	// SplitFailed is the state of a failed split
+	SplitFailed SplitState = "failed"
+
+	// SplitRunning is the state of running split
+	SplitRunning SplitState = "running"
+)
+
+// IsValid checks the value of a diamond state
+func (s SplitState) IsValid() bool {
+	switch s {
+	case SplitDone, SplitFailed, SplitRunning:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s SplitState) String() string {
+	return string(s)
+}
+
+// DiamondDescriptor models a diamond's metadata
+type DiamondDescriptor struct {
+	DiamondID       string            `json:"diamondID" yaml:"diamondID"`
+	StartTime       time.Time         `json:"startTime" yaml:"startTime"` // when the diamond was initialized
+	EndTime         time.Time         `json:"endTime,omitempty" yaml:"endTime,omitempty"`
+	StateChangeTime time.Time         `json:"stateChangeTime" yaml:"stateChangeTime"` // last state change
+	State           DiamondState      `json:"state" yaml:"state"`
+	Mode            ConflictMode      `json:"mode" yaml:"mode"`
+	HasConflicts    bool              `json:"hasConflicts,omitempty" yaml:"hasConflicts,omitempty"`
+	HasCheckpoints  bool              `json:"hasCheckpoints,omitempty" yaml:"hasCheckpoints,omitempty"`
+	Tag             string            `json:"tag,omitempty" yaml:"tag,omitempty"`
```

Comment:
What is tag?

Reply:
Tag is just an informative string to be kept in the logs and in list commands.
I am basically using this to triage log from many splits and coordinator pods

Reply:
It has no impact on the functionality and is essentially here to make log more readable.